[Bug] int4模型triton server请求异常

### Checklist

- [ ] 1. I have searched related issues but cannot get the expected help.
- [ ] 2. The bug has not been fixed in the latest version.

### Describe the bug

使用模型：https://github.com/ymcui/Chinese-LLaMA-Alpaca这个项目开源的Chinese-Alpaca-Pro-13B
部署triton-server后，使用客户端请求：
返回结果直接是这样子的：
![1692258909165](https://github.com/InternLM/lmdeploy/assets/26675984/dcdd0ad0-cdcb-48f5-b940-16a19f27ac6b)


检查triton server日志发现/workspace/models/model_repository/turbomind/1/weights/tok_embeddings.weight has 403548160, but request 403537920, loading model fails!这个有点异常，不清楚正常吗

后面我换成llama2-7b-chat使用int4是可以正常推理的；
使用Chinese-Alpaca-Pro-13B基于fp16也是可以正常推理的；

麻烦检查下啊

### Reproduction
量化步骤：
python3 -m lmdeploy.lite.apis.calibrate --model ../cuiyiming-chinese-alpaca-13b/ --calib_dataset 'c4' --calib_samples 128 --calib_seqlen 2048 --work_dir ./cym-llama-13b-alpaca-int4
python3 -m lmdeploy.lite.apis.auto_awq --model ./cuiyiming-chinese-alpaca-13b/ --w_bits 4 --w_group_size 128 --work_dir ./cym-llama-13b-alpaca-int4
转换成turbomind格式：
python3 -m lmdeploy.serve.turbomind.deploy \
    --model-name llama \
    --model-path ./cym-llama-13b-alpaca-int4 \
    --model-format awq \
    --group-size 128 \
    -d workspace-cym-llama-13b-alpaca-int4
triton-server部署：
bash workspace-cym-llama-13b-alpaca-int4/service_docker_up.sh
trition server部署日志如下：
Triton Server Version 2.29.0

Copyright (c) 2018-2022, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.

Various files include modifications (c) NVIDIA CORPORATION & AFFILIATES.  All rights reserved.

This container image and its contents are governed by the NVIDIA Deep Learning Container License.
By pulling and using the container, you accept the terms and conditions of this license:
https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license

NOTE: CUDA Forward Compatibility mode ENABLED.
  Using CUDA 11.8 driver version 520.61.05 with kernel driver version 510.73.05.
  See https://docs.nvidia.com/deploy/cuda-compatibility/ for details.

I0817 07:36:15.292952 1 pinned_memory_manager.cc:240] Pinned memory pool is created at '0x7f07f8000000' with size 268435456
I0817 07:36:15.293596 1 cuda_memory_manager.cc:105] CUDA memory pool is created on device 0 with size 67108864
I0817 07:36:15.317642 1 model_lifecycle.cc:459] loading: preprocessing:1
I0817 07:36:15.317689 1 model_lifecycle.cc:459] loading: postprocessing:1
I0817 07:36:15.317729 1 model_lifecycle.cc:459] loading: turbomind:1
I0817 07:36:15.530570 1 libfastertransformer.cc:1746] TRITONBACKEND_Initialize: turbomind
I0817 07:36:15.530593 1 libfastertransformer.cc:1753] Triton TRITONBACKEND API version: 1.10
I0817 07:36:15.530598 1 libfastertransformer.cc:1757] 'turbomind' TRITONBACKEND API version: 1.10
I0817 07:36:20.097221 1 libfastertransformer.cc:1784] TRITONBACKEND_ModelInitialize: turbomind (version 1)
I0817 07:36:20.097900 1 libfastertransformer.cc:307] Instance group type: KIND_CPU count: 48
num_nodes=1
tp_pp_size=1
gpu_size=1
world_size=1
model_instance_size=1
I0817 07:36:20.097925 1 libfastertransformer.cc:346] Sequence Batching: disabled
I0817 07:36:20.097929 1 libfastertransformer.cc:357] Dynamic Batching: disabled
[TM][INFO] Barrier(1)
Model:
head_num: 40
kv_head_num: 40
size_per_head: 128
inter_size: 13824
num_layer: 40
vocab_size: 39408
attn_bias: 0
max_batch_size: 32
max_context_token_num: 8224
session_len: 2056
step_length: 1
cache_max_entry_count: 48
cache_chunk_size: 1
use_context_fmha: 1
start_id: 1
tensor_para_size: 1
pipeline_para_size: 1
enable_custom_all_reduce: 0
model_name: llama
model_dir: /workspace/models/model_repository/turbomind/1/weights
quant_policy: 0
group_size: 128
I0817 07:36:20.098636 1 libfastertransformer.cc:387] Before Loading Weights:
after allocation    : free: 23.37 GB, total: 23.69 GB, used:  0.32 GB
[TM][WARNING] file /workspace/models/model_repository/turbomind/1/weights/tok_embeddings.weight has 403548160, but request 403537920, loading model fails!
[TM][WARNING] file /workspace/models/model_repository/turbomind/1/weights/output.weight has 403548160, but request 403537920, loading model fails!
I0817 07:36:24.537778 1 libfastertransformer.cc:396] After Loading Weights:

客户端请求：
python3 -m lmdeploy.serve.client 0.0.0.0:33337 1

### Error traceback

_No response_

> docker image更新好了，重新pull一下 openmmlab/lmdeploy:latest

使用上面的llama1模型验证了下，int4这套方案可以正常推理。
后续为了对比fp16与int4的结果，我又使用了llama2-13b进行了验证：
int4可以正常推理。
fp16部署异常，信息如下：
使用的lmdeploy=0.0.5
转换fp16: python3 -m lmdeploy.serve.turbomind.deploy llama2 ../pretrained-models/llama-2-13b-3e/ hf --tp 2
使用openmmlab/lmdeploy:latest部署triton server
部署日志：
[TM][INFO] Barrier(2)
Model:
head_num: 40
kv_head_num: 40
size_per_head: 128
inter_size: 13824
num_layer: 40
vocab_size: 65165
attn_bias: 0
max_batch_size: 32
max_context_token_num: 16416
session_len: 4104
step_length: 1
cache_max_entry_count: 48
cache_chunk_size: 1
use_context_fmha: 1
start_id: 1
tensor_para_size: 2
pipeline_para_size: 1
enable_custom_all_reduce: 0
model_name: llama2
model_dir: /workspace/models/model_repository/turbomind/1/weights
quant_policy: 0
group_size: 0
I0821 07:36:42.354299 1 libfastertransformer.cc:388] Before Loading Weights:
after allocation    : free: 23.37 GB, total: 23.69 GB, used:  0.32 GB
[TM][WARNING] file /workspace/models/model_repository/turbomind/1/weights/tok_embeddings.weight has 667310080, but request 667289600, loading model fails!
[TM][WARNING] file /workspace/models/model_repository/turbomind/1/weights/tok_embeddings.weight has 667310080, but request 667289600, loading model fails!
[TM][WARNING] file /workspace/models/model_repository/turbomind/1/weights/output.weight has 667310080, but request 667289600, loading model fails!
[TM][WARNING] file /workspace/models/model_repository/turbomind/1/weights/output.weight has 667310080, but request 667289600, loading model fails!
I0821 07:36:52.777463 1 libfastertransformer.cc:397] After Loading Weights:
after allocation    : free: 10.17 GB, total: 23.69 GB, used: 13.52 GB
W0821 07:36:52.777599 1 libfastertransformer.cc:517] skipping model configuration auto-complete for 'turbomind': not supported for turbomind backend
I0821 07:36:52.778134 1 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: preprocessing_0_0 (CPU device 0)
I0821 07:36:53.951062 1 python_be.cc:1856] TRITONBACKEND_ModelInstanceInitialize: postprocessing_0_0 (CPU device 0)
I0821 07:36:55.492449 1 libfastertransformer.cc:422] Before Loading Model:
after allocation    : free: 10.06 GB, total: 23.69 GB, used: 13.63 GB
after allocation    : free: 10.06 GB, total: 23.69 GB, used: 13.63 GB
[WARNING] gemm_config.in is not found; using default GEMM algo
[WARNING] gemm_config.in is not found; using default GEMM algo
terminate called recursively
terminate called after throwing an instance of 'std::runtime_error'
[d319215a1c06:00001] *** Process received signal ***
  what():  [d319215a1c06:00001] Signal: Aborted (6)
[d319215a1c06:00001] Signal code:  (-6)
[TM][ERROR]  Assertion fail: /opt/tritonserver/lmdeploy/src/turbomind/models/llama/LlamaV2.cc:93

[d319215a1c06:00001] [ 0] /usr/lib/x86_64-linux-gnu/libpthread.so.0[d319215a1c06:1    :0:636] Caught signal 11 (Segmentation fault: Sent by the kernel at address (nil))
(+0x14420)[0x7f80bdd24420]
[d319215a1c06:00001] [ 1] /usr/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb)[0x7f80bc5af00b]
[d319215a1c06:00001] [ 2] /usr/lib/x86_64-linux-gnu/libc.so.6(abort+0x12b)[0x7f80bc58e859]
[d319215a1c06:00001] [ 3] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xac53a)[0x7f80bc97653a]
[d319215a1c06:00001] [ 4] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa38c)[0x7f80bc97438c]
[d319215a1c06:00001] [ 5] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa3f7)[0x7f80bc9743f7]
[d319215a1c06:00001] [ 6] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa6a9)[0x7f80bc9746a9]
[d319215a1c06:00001] [ 7] /opt/tritonserver/backends/turbomind/libtransformer-shared.so(_ZN9turbomind17throwRuntimeErrorEPKciRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x23d)[0x7f7ff37e553d]
[d319215a1c06:00001] [ 8] /opt/tritonserver/backends/turbomind/libtransformer-shared.so(_ZN9turbomind7LlamaV2I6__halfEC2EmmmmmmRKNS_20LlamaAttentionParamsEfiiiiiiiiibSt10shared_ptrINS2_11SharedStateEEPNS_11LlamaWeightIS1_EENS_9NcclParamEP11CUstream_stPNS_15cublasMMWrapperEPNS_10IAllocatorEbP14cudaDeviceProp+0x8f4)[0x7f7ff381b0b4]
[d319215a1c06:00001] [ 9] /opt/tritonserver/backends/turbomind/libtransformer-shared.so(_ZN16LlamaTritonModelI6__halfE25createSharedModelInstanceEiiSt4pairISt6vectorIN9turbomind9NcclParamESaIS5_EES7_ESt10shared_ptrINS4_18AbstractCustomCommEE+0x5b2)[0x7f7ff388f5e2]
[d319215a1c06:00001] [10] /opt/tritonserver/backends/turbomind/libtransformer-shared.so(_ZN16LlamaTritonModelI6__halfE19createModelInstanceEiiP11CUstream_stSt4pairISt6vectorIN9turbomind9NcclParamESaIS7_EES9_ESt10shared_ptrINS6_18AbstractCustomCommEE+0x441)[0x7f7ff3894a91]
[d319215a1c06:00001] [11] /opt/tritonserver/backends/turbomind/libtriton_turbomind.so(+0x1548a)[0x7f807028348a]
[d319215a1c06:00001] [12] /opt/tritonserver/backends/turbomind/libtriton_turbomind.so(+0x15d43)[0x7f8070283d43]
[d319215a1c06:00001] [13] /opt/tritonserver/backends/turbomind/libtriton_turbomind.so(+0x32a91)[0x7f80702a0a91]
[d319215a1c06:00001] [14] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xd6de4)[0x7f80bc9a0de4]
[d319215a1c06:00001] [15] /usr/lib/x86_64-linux-gnu/libpthread.so.0(+0x8609)[0x7f80bdd18609]
[d319215a1c06:00001] [16] /usr/lib/x86_64-linux-gnu/libc.so.6(clone+0x43)[0x7f80bc68b133]
[d319215a1c06:00001] *** End of error message ***

检查日志发现，针对llama2的fp16部署还存在2处bug:
1.和最初出现的信息一致：/weights/tok_embeddings.weight has 667310080, but request 667289600, loading model fails!
2.Assertion fail: /opt/tritonserver/lmdeploy/src/turbomind/models/llama/LlamaV2.cc:93

麻烦再审核下啊

--- a/lmdeploy/serve/turbomind/deploy.py
+++ b/lmdeploy/serve/turbomind/deploy.py
@@ -153,8 +153,7 @@ def export(model_name: str,
     if _vocab_size % tp != 0:
         # Resolve https://github.com/InternLM/lmdeploy/issues/266
         # Pad tok_embeddings and output weights, making their shape divisible by TP # noqa: E501
-        _vocab_size = (_vocab_size + tp - 1) // tp * tp
-        pad_size = _vocab_size - tok_embeddings.shape[0]
+        pad_size = (_vocab_size + tp - 1) // tp * tp - _vocab_size
         # Pad weight at the bottom of dim 0
         model_params['tok_embeddings.weight'] = torch.nn.functional.pad(
             tok_embeddings, (0, 0, 0, pad_size), 'constant', 0)

--- a/lmdeploy/serve/turbomind/deploy.py
+++ b/lmdeploy/serve/turbomind/deploy.py
@@ -154,12 +154,13 @@ def export(model_name: str,
         # Resolve https://github.com/InternLM/lmdeploy/issues/266
         # Pad tok_embeddings and output weights, making their shape divisible by TP # noqa: E501
         _vocab_size = (_vocab_size + tp - 1) // tp * tp
+        pad_size = _vocab_size - tok_embeddings.shape[0]
         # Pad weight at the bottom of dim 0
         model_params['tok_embeddings.weight'] = torch.nn.functional.pad(
-            tok_embeddings, (0, 0, 0, 1), 'constant', 0)
+            tok_embeddings, (0, 0, 0, pad_size), 'constant', 0)
         # Pad output weight at the bottom of dim 0
         model_params['output.weight'] = torch.nn.functional.pad(
-            model_params['output.weight'], (0, 0, 0, 1), 'constant', 0)
+            model_params['output.weight'], (0, 0, 0, pad_size), 'constant', 0)
 
     # reverse the splitting axes since the weights are transposed above
     for param_name, param_data in model_params.items():

--- a/lmdeploy/serve/turbomind/deploy.py
+++ b/lmdeploy/serve/turbomind/deploy.py
@@ -147,11 +147,22 @@ def export(model_name: str,
     attn_bias = False
     inter_size = 0
 
+    tok_embeddings = model_params['tok_embeddings.weight']
+    _vocab_size, dim = tok_embeddings.shape
+    head_num = dim // size_per_head
+    if _vocab_size % tp != 0:
+        # Resolve https://github.com/InternLM/lmdeploy/issues/266
+        # Pad tok_embeddings and output weights, making their shape divisible by TP # noqa: E501
+        _vocab_size = (_vocab_size + tp - 1) // tp * tp
+        # Pad weight at the bottom of dim 0
+        model_params['tok_embeddings.weight'] = torch.nn.functional.pad(
+            tok_embeddings, (0, 0, 0, 1), 'constant', 0)
+        # Pad output weight at the bottom of dim 0
+        model_params['output.weight'] = torch.nn.functional.pad(
+            model_params['output.weight'], (0, 0, 0, 1), 'constant', 0)
+
     # reverse the splitting axes since the weights are transposed above
     for param_name, param_data in model_params.items():
-        if param_name == 'tok_embeddings.weight':
-            _vocab_size, dim = param_data.shape
-            head_num = dim // size_per_head
         split_dim = None
         key, ext = param_name.split('.')[-2:]
         if key == 'w_qkv' and ext == 'bias':

--- a/src/turbomind/models/llama/LlamaBatch.cc
+++ b/src/turbomind/models/llama/LlamaBatch.cc
@@ -152,7 +152,7 @@ void LlamaBatch<T>::allocateBuffer(size_t batch_size, size_t session_len)
     const size_t batchxbeam = batch_size;
 
     const size_t hidden_units = llama_->hidden_units_;
-    const size_t vocab_size   = llama_->vocab_size_;
+    const size_t vocab_size   = llama_->vocab_size_padded_;
 
     context_decoder_input_buf_ =
         (T*)allocator_->reMalloc(context_decoder_input_buf_, sizeof(T) * max_context_token_num_ * hidden_units, false);
@@ -899,11 +899,11 @@ void LlamaBatch<T>::outputContextLogits(T*                      context_decoder_
 
     if (context_logits_buf_ == nullptr) {
         NcclGuard guard(llama_->tensor_para_, stream_, true);
-        context_logits_buf_ = (float*)allocator_->malloc(sizeof(float) * llama_->vocab_size_ * max_context_token_num_);
+        context_logits_buf_ = (float*)allocator_->malloc(sizeof(float) * llama_->vocab_size_padded_ * max_context_token_num_);
         const auto tp       = llama_->tensor_para_.world_size_;
         if (tp > 1) {
-            FT_CHECK(llama_->vocab_size_ % tp == 0);
-            const auto local_vocab_size = llama_->vocab_size_ / tp;
+            FT_CHECK(llama_->vocab_size_padded_ % tp == 0);
+            const auto local_vocab_size = llama_->vocab_size_padded_ / tp;
             local_context_logits_buf_ =
                 (float*)allocator_->malloc(sizeof(float) * local_vocab_size * max_context_token_num_);
         }
@@ -921,7 +921,7 @@ void LlamaBatch<T>::outputContextLogits(T*                      context_decoder_
                                              cudaMemcpyDefault,
                                              stream_));
         }
-        logits += llama_->vocab_size_ * lengths[k];
+        logits += llama_->vocab_size_padded_ * lengths[k];
     }
 }
 

--- a/src/turbomind/models/llama/LlamaV2.cc
+++ b/src/turbomind/models/llama/LlamaV2.cc
@@ -71,6 +71,7 @@ LlamaV2<T>::LlamaV2(size_t                       head_num,
     inter_size_(inter_size),
     num_layer_(num_layer),
     vocab_size_(vocab_size),
+    vocab_size_padded_(vocab_size),
     rmsnorm_eps_(norm_eps),
     start_id_(start_id),
     end_id_(end_id),
@@ -90,8 +91,9 @@ LlamaV2<T>::LlamaV2(size_t                       head_num,
 
 {
     TM_LOG_DEBUG(__PRETTY_FUNCTION__);
-    FT_CHECK(vocab_size_ % tensor_para_.world_size_ == 0);
     TM_LOG_INFO("NCCL group_id = %d", tensor_para_.group_id_);
+
+    vocab_size_padded_ = vocab_size_padded_ + vocab_size_padded_ % tensor_para_.world_size_;
 
     size_t elem_bits = 0;
     if (quant_policy & QuantPolicy::kCacheKVInt8) {
@@ -168,7 +170,7 @@ void LlamaV2<T>::initialize(const LlamaAttentionParams& attn_params,
                                    quant_policy);
 
     dynamic_decode_layer_ = new DynamicDecodeLayer<float>(vocab_size_,
-                                                          vocab_size_,  // vocab_size_padded,
+                                                          vocab_size_padded_,
                                                           0,            // end_id, deprecated
                                                           stream_,
                                                           cublas_wrapper_,

--- a/src/turbomind/models/llama/LlamaV2.cc
+++ b/src/turbomind/models/llama/LlamaV2.cc
@@ -93,7 +93,7 @@ LlamaV2<T>::LlamaV2(size_t                       head_num,
     TM_LOG_DEBUG(__PRETTY_FUNCTION__);
     TM_LOG_INFO("NCCL group_id = %d", tensor_para_.group_id_);
 
-    vocab_size_padded_ = vocab_size_padded_ + vocab_size_padded_ % tensor_para_.world_size_;
+    vocab_size_padded_ = (vocab_size_padded_ + tensor_para_.world_size_ - 1) / tensor_para_.world_size_ * tensor_para_.world_size_;
 
     size_t elem_bits = 0;
     if (quant_policy & QuantPolicy::kCacheKVInt8) {

--- a/src/turbomind/models/llama/LlamaV2.cc
+++ b/src/turbomind/models/llama/LlamaV2.cc
@@ -335,8 +335,8 @@ void LlamaV2<T>::postDecodeEmbedding(float* logits, float* local_logits, const T
                               cublasGemmAlgo_t(-1));
     }
     else {
-        FT_CHECK(vocab_size_ % tensor_para_.world_size_ == 0);
-        const size_t local_vocab_size = vocab_size_ / tensor_para_.world_size_;
+        FT_CHECK(vocab_size_padded_ % tensor_para_.world_size_ == 0);
+        const size_t local_vocab_size = vocab_size_padded_ / tensor_para_.world_size_;
         cublas_wrapper_->Gemm(CUBLAS_OP_T,
                               CUBLAS_OP_N,
                               local_vocab_size,  // n
@@ -391,7 +391,7 @@ void LlamaV2<T>::dynamicDecode(int*            token_ids,
     int local_batch_size = (int)batch_size;
 
     std::unordered_map<std::string, Tensor> dynamic_decode_input_tensors{
-        {"logits", {MEMORY_GPU, TYPE_FP32, {batch_size, (size_t)1, vocab_size_}, logits}},
+        {"logits", {MEMORY_GPU, TYPE_FP32, {batch_size, (size_t)1, vocab_size_padded_}, logits}},
         {"step", {MEMORY_CPU, TYPE_INT32, {1}, &step}},
         {"max_input_length", {MEMORY_CPU, TYPE_INT32, {1}, &max_context_len}},
         {"sequence_limit_length", {MEMORY_GPU, TYPE_UINT32, {batch_size}, seq_limit_len}},

--- a/src/turbomind/models/llama/LlamaV2.h
+++ b/src/turbomind/models/llama/LlamaV2.h
@@ -156,7 +156,7 @@ private:
     const size_t inter_size_;
     const size_t num_layer_;
     const size_t vocab_size_;
-    size_t vocab_size_padded_;
+    size_t       vocab_size_padded_;
     float        rmsnorm_eps_ = 1e-6f;
 
     static constexpr bool neox_rotary_style_ = false;

--- a/src/turbomind/models/llama/LlamaV2.h
+++ b/src/turbomind/models/llama/LlamaV2.h
@@ -156,6 +156,7 @@ private:
     const size_t inter_size_;
     const size_t num_layer_;
     const size_t vocab_size_;
+    size_t vocab_size_padded_;
     float        rmsnorm_eps_ = 1e-6f;
 
     static constexpr bool neox_rotary_style_ = false;

--- a/src/turbomind/models/llama/LlamaWeight.cc
+++ b/src/turbomind/models/llama/LlamaWeight.cc
@@ -44,7 +44,7 @@ LlamaWeight<T>::LlamaWeight(size_t     head_num,
     tensor_para_rank_(tensor_para_rank)
 {
     if (vocab_size_padded_ % tensor_para_size_ != 0) {
-        vocab_size_padded_ = vocab_size_padded_ + vocab_size_padded_ % tensor_para_size_;
+        vocab_size_padded_ = (vocab_size_padded_ + tensor_para_size_ - 1) / tensor_para_size_ * tensor_para_size_;
         TM_LOG_WARNING("pad vocab size from %d to %d", vocab_size_, vocab_size_padded_);
     }
     decoder_layer_weights.reserve(num_layer_);

--- a/src/turbomind/models/llama/LlamaWeight.cc
+++ b/src/turbomind/models/llama/LlamaWeight.cc
@@ -37,11 +37,16 @@ LlamaWeight<T>::LlamaWeight(size_t     head_num,
     hidden_units_(head_num * size_per_head),
     inter_size_(inter_size),
     vocab_size_(vocab_size),
+    vocab_size_padded_(vocab_size),
     num_layer_(num_layer),
     weight_type_(weight_type),
     tensor_para_size_(tensor_para_size),
     tensor_para_rank_(tensor_para_rank)
 {
+    if (vocab_size_padded_ % tensor_para_size_ != 0) {
+        vocab_size_padded_ = vocab_size_padded_ + vocab_size_padded_ % tensor_para_size_;
+        TM_LOG_WARNING("pad vocab size from %d to %d", vocab_size_, vocab_size_padded_);
+    }
     decoder_layer_weights.reserve(num_layer_);
     for (unsigned l = 0; l < num_layer_; ++l) {
         decoder_layer_weights.push_back(new LlamaDecoderLayerWeight<T>(head_num,
@@ -72,9 +77,9 @@ LlamaWeight<T>::~LlamaWeight()
 template<typename T>
 void LlamaWeight<T>::mallocWeights()
 {
-    deviceMalloc((T**)&pre_decoder_embedding_table, vocab_size_ * hidden_units_);
+    deviceMalloc((T**)&pre_decoder_embedding_table, vocab_size_padded_ * hidden_units_);
     deviceMalloc((T**)&output_norm_weight, hidden_units_);
-    deviceMalloc((T**)&post_decoder_embedding_kernel, hidden_units_ * vocab_size_);
+    deviceMalloc((T**)&post_decoder_embedding_kernel, hidden_units_ * vocab_size_padded_);
 }
 
 template<typename T>
@@ -84,14 +89,14 @@ void LlamaWeight<T>::loadModel(std::string dir_path)
     dir_path += '/';
 
     loadWeightFromBin((T*)pre_decoder_embedding_table,
-                      {vocab_size_ * hidden_units_},
+                      {vocab_size_padded_ * hidden_units_},
                       dir_path + "tok_embeddings.weight",
                       model_file_type);
 
     loadWeightFromBin((T*)output_norm_weight, {hidden_units_}, dir_path + "norm.weight", model_file_type);
 
     loadWeightFromBin(
-        (T*)post_decoder_embedding_kernel, {hidden_units_ * vocab_size_}, dir_path + "output.weight", model_file_type);
+        (T*)post_decoder_embedding_kernel, {hidden_units_ * vocab_size_padded_}, dir_path + "output.weight", model_file_type);
 
     for (unsigned layer = 0; layer < num_layer_; ++layer) {
         decoder_layer_weights[layer]->loadModel(dir_path + "layers." + std::to_string(layer), model_file_type);

--- a/src/turbomind/models/llama/LlamaWeight.h
+++ b/src/turbomind/models/llama/LlamaWeight.h
@@ -58,6 +58,7 @@ private:
     size_t     hidden_units_;
     size_t     inter_size_;
     size_t     vocab_size_;
+    size_t     vocab_size_padded_;
     size_t     num_layer_;
     WeightType weight_type_;
     size_t     tensor_para_size_;

